### DISCUSSION TITLE
[Rust] Can plugins create `async` http request calls?

### DISCUSSION THREAD

**Discussion by mainrs** [extism/extism#525] [Q&A]
I have a use-case where multiple plugins will do web requests. I want the plugins to be able to spawn multiple simultaneous requests instead of a single request at a time. 

Is it possible to define `async` functions on the WASM side and `await` them on the host system?  If not, what would be the way to do for this? This does not seem to be the case since the `plugin.call` function is sync.

**Comment by nilslice** [comment]
Not across the guest/host boundary yet.. we're able to do this in the JS SDK, if that could work for you.

**Reply by mainrs** [reply]
I am compiling my plugins to Rust's `wasm32-wasi` target. And the host application is also written in Rust. For this setup, this is yet not possible, is it?

Is there any possibility to do this? Or does this rely on features that are just not available in WASM right now?

**Reply by nilslice** [reply]
Sorry, i misread your initial question with mention of:
> I want the plugins to be able to spawn multiple simultaneous requests instead of a single request at a time.

Unfortunately the simply answer is you can't. The extism http functions are synchronous, as they are performed by the host and block to pass the response back to the plugin. 

once WASI preview2 is available more widely, we could look into potentially enabling the plugin to make HTTP calls via WASI, and could do so asynchronously across simultaneous requests.. but that's probably not something that's in the near future. 

you could consider adding your own host function that allows you to take a list of URLs, and your host can make them concurrently for the plugin.. but you'd still have to block in the plugin on the fully awaited responses of each request, and then again from the host on the plugin.
